linkerd/proxy/detect/src/lib.rs

```diff
@@ -90,36 +113,45 @@ where
     }
 }
 
-impl<T, D, A> Future for AcceptFuture<T, D, A>
+impl<T, D, S> Future for DetectProtocolFuture<T, D, S>
 where
     D: Detect<T>,
-    A: core::listen::Accept<(D::Target, BoxedIo)>,
+    S: tower::Service<(D::Target, BoxedIo)>,
+    S::Error: Into<Error>,
 {
-    type Item = ();
+    type Item = S::Response;
     type Error = Error;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
         loop {
             match self {
-                AcceptFuture::Accept(ref mut fut) => return fut.poll().map_err(Into::into),
-                AcceptFuture::Detect {
+                DetectProtocolFuture::Detected(inner_fut) => {
+                    return inner_fut.poll().map_err(Into::into)
+                }
+                DetectProtocolFuture::Detecting {
+                    inner_svc,
                     ref detect,
-                    ref mut accept,
                     ref mut inner,
                 } => match inner {
-                    PeekAndDetect::Peek(ref mut target, ref mut peek) => {
-                        let io = try_ready!(peek.poll().map_err(Error::from));
-                        let target = detect.detect_peeked_prefix(
-                            target.take().expect("polled after complete"),
-                            io.prefix().as_ref(),
-                        );
-                        *inner = PeekAndDetect::Detected(Some((target, BoxedIo::new(io))));
-                    }
                     PeekAndDetect::Detected(ref mut io) => {
-                        try_ready!(accept.poll_ready().map_err(Into::into));
+                        try_ready!(inner_svc.poll_ready().map_err(Into::into));
                         let io = io.take().expect("polled after complete");
-                        *self = AcceptFuture::Accept(accept.accept(io));
+                        let inner_fut = inner_svc.call(io);
+                        *self = DetectProtocolFuture::Detected(inner_fut);
                     }
+                    PeekAndDetect::Peek(ref mut target, ref mut peek) => match peek.poll() {
+                        Ok(Async::NotReady) => return Ok(Async::NotReady),
+                        Ok(Async::Ready(io)) => {
+                            let target = detect.detect_peeked_prefix(
+                                target.take().expect("polled after complete"),
+                                io.prefix().as_ref(),
+                            );
+                            let inner_fut = inner_svc.call((target, BoxedIo::new(io)));
+                            *self = DetectProtocolFuture::Detected(inner_fut);
+                        }
+
+                        Err(err) => return Err(Error::from(err)),
+                    },
```

Comment:
It's not clear to me what's changing here exactly... Would you mind reordering the branches as they were previously (peek, and then detected). It looks like there's some unnecessary changes here... For instance, can't we continue to use `try_ready!(` in the peek branch? Can this diff be made smaller? Is it just renames/reorg or are there functional changes?